app/react/Layout/TableViewer.tsx

```diff
@@ -0,0 +1,63 @@
+import React, { Component } from 'react';
+import { createSelector } from 'reselect';
+
+import { connect } from 'react-redux';
+import { EntityDisplayState, IStore, TableViewColumn } from 'app/istore';
+import { Translate } from 'app/I18N';
+import { TableRows } from 'app/Layout/TableRows';
+import { CollectionViewerProps } from './CollectionViewerProps';
+
+export interface TableViewerProps extends CollectionViewerProps {
+  columns: TableViewColumn[];
+}
+
+class TableViewerComponent extends Component<TableViewerProps> {
+  handleScroll = (e: { target: any }) => {
+    const element = e.target;
+    if (element.scrollHeight - element.scrollTop === element.clientHeight) {
+      this.props.onEndScroll();
+    }
+  };
+
+  render() {
+    const columns = this.props.columns.filter((c: TableViewColumn) => !c.hidden);
+    return (
+      <div className="tableview-wrapper" onScroll={this.handleScroll}>
+        <table>
+          <thead>
+            <tr>
+              {columns.map((column: TableViewColumn, index: number) => (
+                <th className={!index ? 'sticky-col' : ''} key={column.name}>
+                  <div className="table-view-cell">
+                    <Translate>{column.label}</Translate>
+                  </div>
+                </th>
+              ))}
+            </tr>
+          </thead>
+          <tbody>
+            <TableRows
+              {...{
+                columns,
+                clickOnDocument: this.props.clickOnDocument,
+                storeKey: this.props.storeKey,
+                onEndScroll: this.props.onEndScroll,
+              }}
+            />
+          </tbody>
+        </table>
+      </div>
+    );
+  }
+}
+
+const getTableViewColumnsSelector = (state: EntityDisplayState) => state.ui.get('tableViewColumns');
+export const selectTableViewColumns = createSelector(getTableViewColumnsSelector, columns =>
+  columns?.toJS()
+);
+
+const mapStateToProps = (state: IStore, props: TableViewerProps) => ({
+  columns: selectTableViewColumns(state[props.storeKey]),
+});
```

Comment:
I'd like @daneryl to comment on this.  What do you think of using toJS() combined with selectors on mapStateToProps?

Reply:
toJS is always slower than just passing the object through, in this case though this avoids the major bottleneck, that is unnecessary re-renders. if this is a big component rendered only once then this should be fine in terms of performance.